quicklens/src/main/scala-3/com.softwaremill.quicklens/package.scala

```diff
@@ -0,0 +1,290 @@
+      def atOrElse[A](fa: List[A], f: A => A, idx: Int, default: => A): List[A] =
+        fa.updated(idx, f(fa.applyOrElse(idx, Function.const(default))))
+      def index[A](fa: List[A], f: A => A, idx: Int): List[A] =
+        if fa.isDefinedAt(idx) then fa.updated(idx, f(fa(idx))) else fa
+    }
+    given [K, M <: ([V] =>> Map[K, V])] : QuicklensIndexedFunctor[M, K] with {
+      def at[A](fa: M[A], f: A => A, idx: K): M[A] =
+        fa.updated(idx, f(fa(idx))).asInstanceOf[M[A]]
+      def atOrElse[A](fa: M[A], f: A => A, idx: K, default: => A): M[A] =
+        fa.updated(idx, f(fa.applyOrElse(idx, Function.const(default)))).asInstanceOf[M[A]]
+      def index[A](fa: M[A], f: A => A, idx: K): M[A] =
+        if fa.isDefinedAt(idx) then fa.updated(idx, f(fa(idx))).asInstanceOf[M[A]] else fa
+    }
+  }
+
+  extension [F[_]: QuicklensFunctor, A](fa: F[A])
+    def each: A = ???
+    def eachWhere(cond: A => Boolean): A = ???
+
+  extension [F[_]: ([G[_]] =>> QuicklensIndexedFunctor[G, I]), I, A](fa: F[A])
+    def at(idx: I): A = ???
+    def atOrElse(idx: I, default: => A): A = ???
+    def index(idx: I): A = ???
+
+  private val shapeInfo = "Path must have shape: _.field1.field2.each.field3.(...)"
+
+  def toPathModify[S: Type, A: Type](obj: Expr[S], f: Expr[(A => A) => S])(using Quotes): Expr[PathModify[S, A]] = '{ PathModify( ${obj}, ${f} ) }
+
+  def fromPathModify[S: Type, A: Type](pathModify: Expr[PathModify[S, A]])(using Quotes): Expr[(A => A) => S] = '{ ${pathModify}.f }
+
+  def to[T: Type, R: Type](f: Expr[T] => Expr[R])(using Quotes): Expr[T => R] = '{ (x: T) => ${ f('x) } }
+
+  def from[T: Type, R: Type](f: Expr[T => R])(using Quotes): Expr[T] => Expr[R] = (x: Expr[T]) => '{ $f($x) }
+
+  def modifyAllImpl[S, A](obj: Expr[S], focus: Expr[S => A], focusesExpr: Expr[Seq[S => A]])(using qctx: Quotes, tpeS: Type[S], tpeA: Type[A]): Expr[PathModify[S, A]] = {
+    import qctx.reflect.*
+
+    val focuses = focusesExpr match {
+      case Varargs(args) => args
+    }
+
+    val modF1 = fromPathModify(modifyImpl(obj, focus))
+    val modF = to[(A => A), S] { (mod: Expr[A => A]) =>
+      focuses.foldLeft(from[(A => A), S](modF1).apply(mod)) {
+        case (objAcc, focus) =>
+          val modCur = fromPathModify(modifyImpl(objAcc, focus))
+          from[(A => A), S](modCur).apply(mod)
+      }
+    }
+
+    toPathModify(obj, modF)
+  }
+
+  def modifyImpl[S, A](obj: Expr[S], focus: Expr[S => A])(using qctx: Quotes, tpeS: Type[S], tpeA: Type[A]): Expr[PathModify[S, A]] = {
+    import qctx.reflect.*
+
+    enum PathSymbol:
+      case Field(name: String)
+      case FunctionDelegate(name: String, givn: Term, typeTree: TypeTree, args: List[Term])
+
+    def toPath(tree: Tree): Seq[PathSymbol] = {
+      tree match {
+        case Select(deep, ident) =>
+          toPath(deep) :+ PathSymbol.Field(ident)
+        case Apply(Apply(Apply(TypeApply(Ident(s), typeTrees), idents), args), List(ident: Ident)) =>
```

Comment:
I'm guessing this case and the next is to support `.each`, `.at` etc. Could you maybe add some comments, which case corresponds to which usage?

Also, aren't we missing checking that we are only capturing `FunctionDelegate`s for path components that we support, that is: `.each`, `.eachWhere`, `.at` etc. - and not others. I think an `... if ...` like here: https://github.com/softwaremill/quicklens/blob/master/quicklens/src/main/scala-2/com/softwaremill/quicklens/QuicklensMacros.scala#L175-L179 would work?

Reply:
ok